### PR TITLE
feat: pack ProperQF PackedProject into 1 slot (uint128 + uint128)

### PR DESCRIPTION
**Pull request by 0xferit**
## Summary

Packs `ProperQF.PackedProject` from 2 storage slots into 1 (`uint128 + uint128`), halving per-project storage cost:

- **sumContributions**: lossy quantization via `UintQuantizationLib` (shift-32, step = 2^32 ~4.3 nanotoken). Per-project readback is floored; global `totalLinearSum` stays exact via delta cancellation.
- **sumSquareRoots**: lossless `uint128` storage. Ceiling matches the arithmetic constraint (`sumSR^2` must fit `uint256`).

Adds `MIN_VOTE_WEIGHT = 2^16` enforcement in `QuadraticVotingMechanism` to guarantee all contributions are step-aligned, preventing silent quantization loss.

The quantization scheme is calibrated for 18-decimal-normalized voting power. `ProperQF` constructor now requires `inputDecimals = 18`, enforced at deploy time, to prevent misuse with lower-precision assets where the step size would silently swallow contributions.

## Gas savings

Measured against develop with identical benchmark workload (1000 projects, 3950 votes, with votes concentrated in a few popular projects and sparse across most others):

| Scenario | Old (2-slot) | New (1-slot) | Difference |
|----------|-------------|-------------|---------|
| Full epoch lifetime | 67.8M | 50.9M | **-24.9%** |
| Seed 1000 projects (cold writes) | 49.2M | 28.4M | -42.2% |
| Tally 1000 projects (reads) | 52.3M | 31.9M | -39.0% |
| Single vote (cold) | 123,611 | 102,881 | -16.8% |
| Single vote (warm) | 128,750 | 109,190 | -15.2% |

~20.7k gas saved per cold project write, ~19.6k per warm write, ~20.4k per read. Each corresponds to exactly 1 eliminated SSTORE/SLOAD.

## Why this is safe

The lossy quantization of `sumContributions` does not affect funding outcomes because:

1. **Per-project loss is bounded and negligible**: floor-rounding loses at most 1 step (~4.3 nanotoken) per vote. This is economically irrelevant at 18-decimal precision.
2. **Global sums stay exact**: `totalLinearSum` and `totalQuadraticSum` are never read from packed storage. The delta update pattern in `_processVoteUnchecked` computes `totalLinearSum = totalLinearSum - decoded_old + decoded_new`, so quantization errors cancel algebraically. Formally verified by P3 and P10.
3. **Alignment is enforced at the entry point**: `QuadraticVotingMechanism` requires `weight >= MIN_VOTE_WEIGHT` and `weight % MIN_VOTE_WEIGHT == 0`. Since `contribution = weight^2` and `MIN_VOTE_WEIGHT^2 == stepSize`, every contribution is step-aligned. This means per-project readback is actually lossless on all reachable paths. Formally verified by P1 and P5.
4. **Overflow is impossible on reachable paths**: `sumSquareRoots` fits in `uint128` because `sumSR^2` must fit in `uint256` (formally verified by P4). `sumContributions` fits in the quantized `uint128` with massive headroom: over 4 billion max-size votes per project before overflow (formally verified by P6 and P11).
5. **Packed halves do not interfere**: the two `uint128` fields in a single slot are independent; writing one does not corrupt the other. Formally verified by P9.

## Kontrol formal verification

12 proofs (all passing) covering quantization invariants, packed storage independence, end-to-end `_processVoteUnchecked` correctness, production/harness drift guards, and live-cap headroom. K lemmas added for `#buf`/`#asWord` simplification of shift-32 encode/decode.

## Known issue

`AllocationMechanismFactory` exceeds EIP-170 by 140 bytes due to increased `ProperQF` bytecode inlined in the factory's creation code. Requires factory pattern change (clone/proxy); will be addressed in a follow-up. The CI Build job fails on this check.

### CODE REVIEW

**Comment by 0xferit** [comment]
@codex review

File: src/mechanisms/voting-strategy/ProperQF.sol [hunk 1]
```diff
@@ -201,18 +245,37 @@ abstract contract ProperQF {
...
+        sumContributions = CONTRIBUTIONS_SCHEME.decode(uint256(packed.sumContributions));
+        sumSquareRoots = SQUARE_ROOTS_SCHEME.decode(uint256(packed.sumSquareRoots));
+    }
+
+    /// @notice Encode and store full-width values into packed project storage
+    /// @dev Reverts with Overflow(value, max) if either field exceeds its scheme's max.
+    ///      The unchecked downcasts are safe because encode() guarantees the result fits.
+    function _writeProject(uint256 projectId, uint256 sumContributions, uint256 sumSquareRoots) internal {
+        _getProperQFStorage().projects[projectId] = PackedProject({
+            sumContributions: uint96(CONTRIBUTIONS_SCHEME.encode(sumContributions)),
```
[the remark this answers is not in the document]
Comment (0xferit) [reply src/mechanisms/voting-strategy/ProperQF.sol]: Both this and the sibling finding (silent zeroing for small weights) are addressed by the same fix: a `ContributionBelowMinimum` guard in `_processVoteUnchecked` that rejects `contribution < CONTRIBUTIONS_SCHEME.stepSize()` (= 2^32).

**Why remainders don't compound:** `_readProject` always returns a step-aligned value (stored << 32), so the floor loss per vote is strictly `contribution mod STEP`, independent of accumulated history. Remainders from previous votes have already been truncated and don't interact with new contributions.

**Bounded loss after the guard:** With `contribution >= STEP` enforced, the per-vote loss ratio is `remainder / contribution`. For the minimum weight (65536), contribution = 2^32 = STEP exactly, so loss = 0. For realistic weights the loss ratio is negligible (e.g. weight=1e9: ~5e-8% per vote).

**Global sums stay exact:** `totalLinearSum` uses the delta pattern `old_decoded - old_decoded + (old_decoded + contribution)`, which cancels the quantization error. Only per-project readback in `getTally` is lossy, bounded by `N * (STEP - 1)` across N votes.

Preserving the remainder would require either an extra storage slot per project (defeating the packing goal) or a fundamentally different accumulation strategy. Given the guard makes the worst case precision dust, I think the tradeoff is right.

**Comment by 0xferit** [comment]
@codex review

File: src/mechanisms/voting-strategy/ProperQF.sol [hunk 2]
```diff
@@ -201,18 +265,37 @@ abstract contract ProperQF {
...
         );
     }
 
+    // ── Pack/unpack helpers ──────────────────────────────────────────────
+
+    /// @notice Decode packed project storage into full-width uint256 values
+    function _readProject(uint256 projectId) internal view returns (uint256 sumContributions, uint256 sumSquareRoots) {
+        PackedProject storage packed = _getProperQFStorage().projects[projectId];
+        sumContributions = CONTRIBUTIONS_SCHEME.decode(uint256(packed.sumContributions));
+        sumSquareRoots = uint256(packed.sumSquareRoots);
```
[the remark this answers is not in the document]
Comment (0xferit) [reply src/mechanisms/voting-strategy/ProperQF.sol:297]: Known issue and won't fix.

File: test/unit/mechanisms/allocation/RecipientVerificationTest.sol [hunk 3]
```diff
@@ -325,9 +326,9 @@ contract RecipientVerificationTest is Test {
         _tokenized(address(mechanism)).castVote(pid1, TokenizedAllocationMechanism.VoteType.For, 5, recipientB);
```
[the remark this answers is not in the document]
Comment (0xferit) [reply test/unit/mechanisms/allocation/RecipientVerificationTest.sol:326]: Not a bug. RecipientMismatch is checked at TokenizedAllocationMechanism.sol:895, before the vote flow reaches QVM's weight checks at line 221. The weight value is irrelevant here because the test asserts RecipientMismatch, which fires first in the call stack. Same for line 289 (VoteType.Against hits OnlyForVotesSupported before any weight check) and line 364.

File: test/unit/mechanisms/allocation/QuadraticVotingMechanism/ProperQFOverflowTest.sol [hunk 4]
```diff
@@ -16,167 +17,151 @@ contract ProperQFHarness is ProperQF {
...
-    /// @notice Test overflow boundary - find the actual overflow point
-    function testOverflowBoundary() public pure {
-        uint256 maxWeight = 22;
-        uint256 contribution = maxWeight * maxWeight;
-        uint256 uint128Max = type(uint128).max;
+    /// @notice Test the overflow boundary for sumContributions
+    function testOverflowBoundary_Contributions() public pure {
+        console.log("=== CONTRIBUTIONS OVERFLOW BOUNDARY ===");
+        console.log("CONTRIBUTIONS_MAX:", CONTRIBUTIONS_MAX);
+        console.log("STEP:", STEP);
```
[the remark this answers is not in the document]
Comment (0xferit) [reply test/unit/mechanisms/allocation/QuadraticVotingMechanism/ProperQFOverflowTest.sol:92]: Not a real issue. Forge treats console.log as a cheatcode that compiles and runs in any mutability context. The test passes as-is.

File: test/unit/mechanisms/allocation/QuadraticVotingMechanism/ProperQFOverflowTest.sol [hunk 5]
```diff
@@ -16,167 +17,151 @@ contract ProperQFHarness is ProperQF {
...
-        properQF.processVoteUnchecked(projectId, hugeContrib, hugeWeight);
+    /// @notice Sub-step contribution reverts with BelowMinStep from UintQuantizationLib
+    function testSubStepEncodesToZero() public {
+        uint256 projectId = 1;
+        // Any value < STEP is rejected by the minimum contribution check
+        uint256 tinyContribution = STEP - 1;
 
-        console.log("SUCCESS: Overflow protection working correctly");
+        vm.expectRevert(abi.encodeWithSelector(BelowMinStep.selector, tinyContribution, STEP));
+        properQF.processVoteUnchecked(projectId, tinyContribution, 1);
```
[the remark this answers is not in the document]
Comment (0xferit) [reply test/unit/mechanisms/allocation/QuadraticVotingMechanism/ProperQFOverflowTest.sol:165]: Good catch. Renamed to testSubStepRevertsWithBelowMinStep in 12ab7abe.

**Comment by 0xferit** [comment]
@codex review

File: src/mechanisms/voting-strategy/ProperQF.sol [hunk 6]
```diff
@@ -10,10 +11,56 @@ import { Math } from "@openzeppelin/contracts/utils/math/Math.sol";
...
+    /// @dev Contributions are quadratic costs (weight^2) in 18-decimal-normalized voting-power
+    ///      space (see QuadraticVotingMechanism._normalizeToDecimals), so these bounds hold
+    ///      uniformly regardless of the underlying asset's native decimals.
+    ///      Step size: 2^32 ~= 4.3e9 wei (~4.3 nanotoken) : floor-rounding error per encode.
+    ///      Max value: (2^128 - 1) << 32 ~= 1.46e48 wei (~1.46e30 tokens).
+    ///      Global sums (totalLinearSum) stay exact because quantization errors cancel in the
+    ///      delta update pattern of _processVoteUnchecked; only per-project readback is lossy.
+    /// Equivalent to QuantLib.create(32, 128); constant required to avoid immutable bloat.
+    /// Cross-validated by Kontrol proof P7 (testProductionSchemeConsistency).
+    Quant internal constant CONTRIBUTIONS_SCHEME = Quant.wrap(uint16((128 << 8) | 32));
```
[the remark this answers is not in the document]
Comment (0xferit) [reply src/mechanisms/voting-strategy/ProperQF.sol:58]: Known issue; wont fix.

File: src/mechanisms/voting-strategy/ProperQF.sol [hunk 7]
```diff
@@ -10,10 +11,58 @@ import { Math } from "@openzeppelin/contracts/utils/math/Math.sol";
  * @notice Incremental QF tallying utilities with alpha-weighted quadratic/linear funding.
  * @dev Provides storage isolation via deterministic slot, input validation helpers,
  *      and funding aggregation with well-defined rounding behavior.
+ *
+ *      Storage packing: each project occupies exactly 1 slot (256 bits):
+ *        sumContributions : uint128, shift=32 (lossy quantization via UintQuantizationLib)
```
Comment (0xferit) [inline src/mechanisms/voting-strategy/ProperQF.sol:16]: Lossless actually, we are enforcing precision. We could also not enforce precision and accept losing the dust. Both is fine.
